dataclients/kubernetes/clusterclient.go

```diff
@@ -449,6 +459,88 @@ func (c *clusterClient) loadEndpoints() (map[definitions.ResourceID]*endpoint, e
 	return result, nil
 }
 
+// loadEndpointSlices is different from the other load$Kind()
+// functions because there are 1..N endpointslices created for a given
+// service. endpointslices need to be deduplicated and state needs to
+// be checked. We read all endpointslices and create de-duplicated
+// business objects skipperEndpointSlice instead of raw Kubernetes
+// objects, because we need just a clean list of load balancer
+// members. The returned map will return the full list of ready
+// non-terminating endpoints that should be in the load balancer of a
+// given service, check endpointSlice.ToResourceID().
+func (c *clusterClient) loadEndpointSlices() (map[definitions.ResourceID]*skipperEndpointSlice, error) {
+	var endpointSlices endpointSliceList
+
+	if err := c.getJSON(c.endpointSlicesURI+c.endpointSlicesLabelSelectors, &endpointSlices); err != nil {
+		log.Debugf("requesting all endpointslices failed: %v", err)
+		return nil, err
+	}
+
+	log.Debugf("all endpointslices received: %d", len(endpointSlices.Items))
+	mapSlices := make(map[definitions.ResourceID][]*endpointSlice)
+	for _, endpointSlice := range endpointSlices.Items {
+		resID := endpointSlice.ToResourceID() // service resource ID
+		mapSlices[resID] = append(mapSlices[resID], endpointSlice)
+	}
+
+	result := make(map[definitions.ResourceID]*skipperEndpointSlice)
+	for resID, epSlices := range mapSlices {
+		if len(epSlices) == 0 {
+			continue
+		}
+
+		result[resID] = &skipperEndpointSlice{
+			Meta: epSlices[0].Meta,
+		}
+
+		for i := range epSlices {
+			resEps := make(map[string]*skipperEndpoint)
+			terminatingEps := make(map[string]struct{})
+
+			for _, ep := range epSlices[i].Endpoints {
+				// if conditions are nil then we need to treat is as ready
+				if ep.Conditions == nil {
+					resEps[ep.Addresses[0]] = &skipperEndpoint{
+						// Addresses [1..100] of the same AddressType, as kube-proxy we use the first
+						// see also https://github.com/kubernetes/kubernetes/issues/106267
+						Address: ep.Addresses[0],
+						Zone:    ep.Zone,
+					}
+					continue
+				}
+
+				if ep.isTerminating() {
+					terminatingEps[ep.Addresses[0]] = struct{}{}
+					// if we had this one with a non termiating condition, we should delete it, because of eventual consistency it is actually terminating
+					delete(resEps, ep.Addresses[0])
+					continue
+				}
+				// already known terminating?
+				if _, ok := terminatingEps[ep.Addresses[0]]; ok {
+					continue
+				}
```

Comment:
Does this mean it was observed termination but now it is not (because previous ep.isTerminating() was false)?

Reply:
Should we log a warning maybe?

Reply:
We do not want to have endpoints with `Terminating: true` in the list of load balancer members.
I don't expect a terminating pods to become `Terminating: false` again.
Endpointslices can have duplicate entries with different states, so if you have 2 slices and one endpoint in both (think also eventual consistency), it should be enough to know that one is terminating to drop it out of the member list.

Reply:
I had to change the order of the logic based on testing input.